inventory-api/src/main/scala/com/normation/inventory/domain/AgentTypes.scala

```diff
@@ -101,43 +103,94 @@ object AgentType {
+          case None => Failure("No value defined for security token")
+        }
+        case JString(PublicKey.kind) => Failure("Cannot have a public Key for dsc agent, only a certificate is valid")
+        case JNothing => Failure("No value define for security token")
+        case invalidJson => Failure(s"Invalid value for security token, ${compactRender(invalidJson)}")
+      }
+      case _ => tokenJson \ "type" match {
+        case JString(Certificate.kind) => extractValue(tokenJson) match {
+          case Some(token) => Full(Certificate(token))
+          case None => Failure("No value defined for security token")
+        }
+        case JString(PublicKey.kind) => extractValue(tokenJson) match {
+          case Some(token) => Full(PublicKey(token))
+          case None => Failure("No value defined for security token")
+        }
+        case invalidJson =>
+          tokenDefault match {
+            case Some(default) => Full(PublicKey(default))
+            case None => Failure(s"Invalid value for security token, ${compactRender(invalidJson)}, and no public key were stored")
+          }
+      }
+    }
   }
 
   /*
    * Retrieve the agent information from JSON. "agentType" is mandatory,
    * but version isn't, and even if we don't parse it correctly, we
    * successfully return an agent (without version).
    */
-  def parseJson(s: String): Box[AgentInfo] = {
+  def parseJson(s: String, optToken : Option[String]): Box[AgentInfo] = {
     for {
       json <- try { Full(parse(s)) } catch { case ex: Exception => Failure(s"Can not parse agent info: ${ex.getMessage }", Full(ex), Empty) }
-      info <- (json \ "agentType") match {
-                case JString(tpe) => AgentType.fromValue(tpe).map { agentType =>
-                  (json \ "version") match {
-                    case JString(version) => AgentInfo(agentType, Some(AgentVersion(version)))
-                    case _                => AgentInfo(agentType, None)
-                  }
-                }
-                case _ => Failure(s"Error when trying to parse string as JSON Agent Info (missing required field 'agentType'): ${s}")
+      agentType <- (json \ "agentType") match {
+                     case JString(tpe) => AgentType.fromValue(tpe)
+                     case JNothing => Failure("No value defined for security token")
+                     case invalidJson  => Failure(s"Error when trying to parse string as JSON Agent Info (missing required field 'agentType'): ${compactRender(invalidJson)}")
+                   }
+     agentVersion = json \ "version" match {
+                      case JString(version) => Some(AgentVersion(version))
+                      case _                => None
+                    }
+     token <- json \ "securityToken" match {
+                case JObject(json) => parseSecurityToken(agentType, json, optToken)
+                case _             => parseSecurityToken(agentType, JNothing, optToken)
```

Comment:
since tokenJson \ "type" will be invalid in this case, this will always return Failure. And fail